File: EmojiTapper/Shared/Game/ClassicGameEngine.swift
//
//  ClassicGameEngine.swift
//  Shared
//
//  Created by Josh Estelle on 8/8/25.
//

import SwiftUI
import Foundation

@Observable
class ClassicGameEngine: GameModeEngine {
    let gameMode: GameMode = .classic
    var score: Int = 0
    var timeRemaining: TimeInterval = 10.0
    var isGameActive: Bool = false
    var currentEmojis: [GameEmoji] = []
    var currentLevel: GameLevel = BasicLevel()
    var highScore: Int = 0 {
        didSet {
            saveHighScore()
        }
    }
    
    var gameStateText: String {
        return String(format: "%.1f", timeRemaining)
    }
    
    // For progress bar in UI
    var timeRemainingForProgress: TimeInterval { timeRemaining }
    var totalTimeForProgress: TimeInterval { currentLevel.initialTime }
    var onEmojisChanged: (() -> Void)?
    var onGameEnded: (() -> Void)?
    
    private var gameTimer: Timer?
    
    private let skullEmoji = GameEmoji(emoji: "💀", type: .skull)
    private let hourglassEmoji = GameEmoji(emoji: "⏳", type: .hourglass)
    private let cherryEmoji = GameEmoji(emoji: "🍒", type: .cherry)
    
    init() {
        loadHighScore()
    }
    
    private var maxEmojisOnScreen: Int {
        let elapsed = currentLevel.initialTime - timeRemaining
        // Start with 1 emoji, add 1 more every 2 seconds, max of 50
        return min(50, 1 + Int(elapsed / 2.0))
    }
    
    func startGame() {
        score = 0
        timeRemaining = currentLevel.initialTime
        isGameActive = true
        generateNewEmojis()
        startTimer()
    }
    
    func endGame() {
        isGameActive = false
        stopTimer()
        
        if score > highScore {
            highScore = score
        }
        
        // Notify that game has ended
        onGameEnded?()
    }
    
    func emojiTapped(_ emoji: GameEmoji) {
        guard isGameActive else { return }
        
        switch emoji.type {
        case .normal:
            let points = currentLevel.pointsForTapping(emoji.emoji)
            score += points
            if points > 0 {
                timeRemaining += timeRemaining * currentLevel.timeBonus
            }
        case .skull:
            endGame()
            return
        case .hourglass:
            timeRemaining += 2.0
        case .cherry:
            score += 3
        }
        
        generateNewEmojis()
        onEmojisChanged?()
    }
    
    func proceedToNextRound() {
        // Classic mode doesn't have rounds, so this does nothing
    }
    
    func pauseTimers() {
        // Classic mode doesn't need timer pausing for position animations
    }
    
    func resumeTimers() {
        // Classic mode doesn't need timer resuming for position animations
    }
    
    func generateNewEmojis() {
        currentEmojis.removeAll()
        let targetCount = maxEmojisOnScreen
        
        // Always add at least one normal emoji
        if let (emoji, type) = EmojiProvider.classicNormal.randomElement() {
            currentEmojis.append(GameEmoji(emoji: emoji, type: type))
        }
        
        // Fill remaining slots
        let remainingSlots = targetCount - currentEmojis.count
        for _ in 0..<remainingSlots {
            let randomValue = Double.random(in: 0...1)
            
            if randomValue < 0.4 {
                // Add skull (40% chance)
                currentEmojis.append(skullEmoji)
            } else if randomValue < 0.5 {
                // Add hourglass (10% chance, max 1)
                if !currentEmojis.contains(where: { $0.type == .hourglass }) {
                    currentEmojis.append(hourglassEmoji)
                } else if let (emoji, type) = EmojiProvider.classicNormal.randomElement() {
                    currentEmojis.append(GameEmoji(emoji: emoji, type: type))
                }
            } else if randomValue < 0.6 {
                // Add cherry (10% chance, max 3)
                if currentEmojis.filter({ $0.type == .cherry }).count < 3 {
                    currentEmojis.append(cherryEmoji)
                } else if let (emoji, type) = EmojiProvider.classicNormal.randomElement() {
                    currentEmojis.append(GameEmoji(emoji: emoji, type: type))
                }
            } else {
                // Add normal emoji
                if let (emoji, type) = EmojiProvider.classicNormal.randomElement() {
                    currentEmojis.append(GameEmoji(emoji: emoji, type: type))
                }
            }
        }
    }
    
    private func startTimer() {
        gameTimer = Timer.scheduledTimer(withTimeInterval: 0.1, repeats: true) { _ in
            self.timeRemaining -= 0.1
            if self.timeRemaining <= 0 {
                self.endGame()
            }
        }
    }
    
    private func stopTimer() {
        gameTimer?.invalidate()
        gameTimer = nil
    }
    
    private func saveHighScore() {
        UserDefaults.standard.set(highScore, forKey: "EmojiTapperClassicHighScore")
    }
    
    private func loadHighScore() {
        highScore = UserDefaults.standard.integer(forKey: "EmojiTapperClassicHighScore")
    }
    
    func resetHighScore() {
        highScore = 0
        UserDefaults.standard.set(0, forKey: "EmojiTapperClassicHighScore")
    }
}

File: EmojiTapperAndroid/app/src/main/java/com/emojitapper/game/ClassicGameEngine.kt
package com.emojitapper.game

import android.content.Context
import android.content.SharedPreferences
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableDoubleStateOf
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableStateListOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.emojitapper.models.*
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlin.math.min

class ClassicGameEngine(private val context: Context) : ViewModel(), GameModeEngine {
    override val gameMode = GameMode.CLASSIC
    
    override var score by mutableIntStateOf(0)
    override var isGameActive by mutableStateOf(false)
    override val currentEmojis = mutableStateListOf<GameEmoji>()
    
    private var timeRemaining by mutableDoubleStateOf(10.0)
    override val timeRemainingForProgress: Double get() = timeRemaining
    override val totalTimeForProgress: Double get() = currentLevel.initialTime
    
    override val gameStateText: String
        get() = String.format("%.1f", timeRemaining)
    
    override var highScore by mutableIntStateOf(0)
        private set
    
    override var onEmojisChanged: (() -> Unit)? = null
    override var onGameEnded: (() -> Unit)? = null
    
    private var currentLevel: GameLevel = BasicLevel()
    private var gameTimer: Job? = null
    
    private val normalEmojis = listOf(
        "😀", "😊", "😂", "🥰", "😎", "🤔", "😮", "😋", "🙂", 
        "😆", "😍", "🤗", "😴", "🤯", "😇"
    )
    private val specialEmojis = listOf("💀", "⏳", "🍒")
    
    private val prefs: SharedPreferences = context.getSharedPreferences(
        "EmojiTapperPrefs", Context.MODE_PRIVATE
    )
    
    init {
        loadHighScore()
    }
    
    private val maxEmojisOnScreen: Int
        get() {
            val elapsed = currentLevel.initialTime - timeRemaining
            return min(50, 1 + (elapsed / 2.0).toInt())
        }
    
    override fun startGame() {
        score = 0
        timeRemaining = currentLevel.initialTime
        isGameActive = true
        generateNewEmojis()
        startTimer()
    }
    
    override fun endGame() {
        isGameActive = false
        stopTimer()
        
        if (score > highScore) {
            highScore = score
            saveHighScore()
        }
        
        onGameEnded?.invoke()
    }
    
    override fun emojiTapped(emoji: GameEmoji) {
        if (!isGameActive) return
        
        when (emoji.type) {
            EmojiType.NORMAL -> {
                val points = currentLevel.pointsForTapping(emoji.emoji)
                score += points
                if (points > 0) {
                    timeRemaining += timeRemaining * currentLevel.timeBonus
                }
            }
            EmojiType.SKULL -> {
                endGame()
                return
            }
            EmojiType.HOURGLASS -> {
                timeRemaining += 1.0
            }
            EmojiType.CHERRY -> {
                score += 2
            }
        }
        
        generateNewEmojis()
        onEmojisChanged?.invoke()
    }
    
    override fun proceedToNextRound() {
        // Classic mode doesn't have rounds
    }
    
    override fun pauseTimers() {
        // Classic mode doesn't need timer pausing
    }
    
    override fun resumeTimers() {
        // Classic mode doesn't need timer resuming
    }
    
    override fun resetHighScore() {
        highScore = 0
        prefs.edit().putInt("EmojiTapperClassicHighScore", 0).apply()
    }
    
    private fun generateNewEmojis() {
        currentEmojis.clear()
        val targetCount = maxEmojisOnScreen
        
        var zIndex = 0
        
        // Always add at least one normal emoji
        val normalEmoji = normalEmojis.random()
        currentEmojis.add(
            GameEmoji(
                emoji = normalEmoji,
                type = EmojiType.NORMAL,
                zIndex = zIndex++
            )
        )
        
        // Fill remaining slots
        val remainingSlots = targetCount - 1
        repeat(remainingSlots) {
            val randomValue = kotlin.random.Random.nextDouble()
            
            when {
                randomValue < 0.4 -> {
                    // Add skull (40% chance)
                    currentEmojis.add(
                        GameEmoji(
                            emoji = "💀",
                            type = EmojiType.SKULL,
                            zIndex = zIndex++
                        )
                    )
                }
                randomValue < 0.5 -> {
                    // Add hourglass if we don't have one (10% chance)
                    val hasHourglass = currentEmojis.any { it.type == EmojiType.HOURGLASS }
                    if (!hasHourglass) {
                        currentEmojis.add(
                            GameEmoji(
                                emoji = "⏳",
                                type = EmojiType.HOURGLASS,
                                zIndex = zIndex++
                            )
                        )
                    } else {
                        // Add normal emoji instead
                        currentEmojis.add(
                            GameEmoji(
                                emoji = normalEmojis.random(),
                                type = EmojiType.NORMAL,
                                zIndex = zIndex++
                            )
                        )
                    }
                }
                randomValue < 0.6 -> {
                    // Add cherry if we have less than 3 (10% chance)
                    val cherryCount = currentEmojis.count { it.type == EmojiType.CHERRY }
                    if (cherryCount < 3) {
                        currentEmojis.add(
                            GameEmoji(
                                emoji = "🍒",
                                type = EmojiType.CHERRY,
                                zIndex = zIndex++
                            )
                        )
                    } else {
                        // Add normal emoji instead
                        currentEmojis.add(
                            GameEmoji(
                                emoji = normalEmojis.random(),
                                type = EmojiType.NORMAL,
                                zIndex = zIndex++
                            )
                        )
                    }
                }
                else -> {
                    // Add normal emoji (40% chance)
                    currentEmojis.add(
                        GameEmoji(
                            emoji = normalEmojis.random(),
                            type = EmojiType.NORMAL,
                            zIndex = zIndex++
                        )
                    )
                }
            }
        }
    }
    
    private fun startTimer() {
        gameTimer = viewModelScope.launch {
            while (isGameActive && timeRemaining > 0) {
                delay(100) // 0.1 second intervals
                timeRemaining -= 0.1
                if (timeRemaining <= 0) {
                    endGame()
                }
            }
        }
    }
    
    private fun stopTimer() {
        gameTimer?.cancel()
        gameTimer = null
    }
    
    private fun saveHighScore() {
        prefs.edit().putInt("EmojiTapperClassicHighScore", highScore).apply()
    }
    
    private fun loadHighScore() {
        highScore = prefs.getInt("EmojiTapperClassicHighScore", 0)
    }
    
    override fun onCleared() {
        super.onCleared()
        stopTimer()
    }
}